Initial Setup for Transaction Storage System

YoussefmSaber opened the pull request:
# Summary :
This pull request contains initial setup for transaction storage system. It covers the file structure, storage implementations, and utility functions for handling transaction.

## Changes :
- Created the base test functions and test cases for JSON utilities and file storage.
- Created base classes for file-based and in-memory transaction storage.
- Implemented serialization and deserialization for transactions.
- Added the DataSource interface for both storage types.
- Added TransactionHolder to manage transactions from both sources and have a single object across all operations.
- Wrote basic documentation for all classes and functions.
- Created the base test functions and test cases for JSON utilities and file storage.
## Screenshots for file structure : 
- Project structure
![image](https://github.com/user-attachments/assets/2298537e-c405-49c7-8278-73c89dcfc3db)
- test structure
![image](https://github.com/user-attachments/assets/dc5d349c-bfc8-4f08-97d0-42620bdc3dd7)

YoussefmSaber committed:
created the base file structure for the porject

YoussefmSaber committed:
created empty classes to implement test cases

YoussefmSaber committed:
created the base test functions for the JsonUtilTest

YoussefmSaber committed:
created the base test cases for FileTransactionStorageTest.kt

YoussefmSaber committed:
created the base class for testing InMemeoryTransactionStorage

YoussefmSaber committed:
created the base files for storing the data in text file

YoussefmSaber committed:
created the base code for serialize and deserialize for transaction object

YoussefmSaber committed:
created the base function for serializeTransactionList

YoussefmSaber committed:
created the base function for deserializeTransactionList

YoussefmSaber committed:
created the interface for data source

YoussefmSaber committed:
made fileTransactionStorage, InMemoryTransactionStorage inherit the data source interface

YoussefmSaber committed:
implemented the logic for saveTransactions function in the FileTransactionStorage

YoussefmSaber committed:
implemented the logic for saveTransaction function

YoussefmSaber committed:
implemented the logic for loadTransactions function from the storage

YoussefmSaber committed:
implemented the functinality for InMemoryTransactionStorage

YoussefmSaber committed:
added TransactionHolder to manage transactions in memory and file storage

YoussefmSaber committed:
enhanced FileTransactionStorage with detailed documentation for transaction saving and loading methods

YoussefmSaber committed:
enhanced InMemoryTransactionStorage with comprehensive documentation for transaction methods

YoussefmSaber committed:
enhanced DataSource interface with detailed documentation for transaction methods

YoussefmSaber committed:
enhanced JsonUtil with serialization and deserialization methods for transactions

YoussefmSaber committed:
refactored TransactionHolder to load transactions from memory instead of file storage

YoussefmSaber committed:
modified the code to use transaction data class

YoussefmSaber committed:
modified the gitignore file

YoussefmSaber committed:
modified the gitignore file

YoussefmSaber committed:
modified the gitignore file

ay0o5h commented on app/src/main/java/com/personalfinancetracer/repository/DataSource.kt:
```diff
@@ -0,0 +1,29 @@
+package com.personalfinancetracer.repository
```
we will not need a repository will keep things simple for now and but this file in data folder and rename data folder to datasource

ay0o5h commented on app/src/main/java/com/personalfinancetracer/repository/DataSource.kt:
```diff
@@ -0,0 +1,29 @@
+     * @param transactions The list of transactions to be saved.
+     * @return `true` if all transactions were saved successfully, `false` otherwise.
+     */
+    fun saveTransactions(transactions: List<Transaction>): Boolean
```
we will create only one transaction at time no need to store list of transiction

YoussefmSaber replied:
ok on it

ay0o5h commented on app/src/main/java/com/personalfinancetracer/data/storage/FileTransactionStorage.kt:
```diff
@@ -0,0 +1,76 @@
+     * @param transactions The list of transactions to save.
+     * @return `true` if the transactions were successfully saved, `false` otherwise.
+     */
+    override fun saveTransactions(transactions: List<Transaction>): Boolean {
```
save only one transiction at time

ay0o5h commented on app/src/main/java/com/personalfinancetracer/repository/DataSource.kt:
```diff
@@ -0,0 +1,29 @@
+     *
+     * @return A list of transactions.
+     */
+    fun loadTransactions(): List<Transaction>
```
rename function to get all transiction

ay0o5h replied:
also add edit and delete transiction

YoussefmSaber replied:
ok

ay0o5h commented on app/src/main/java/com/personalfinancetracer/data/storage/TransactionHolder.kt:
since we committed to file-based storage, you can remove InMemoryTransactionStorage and TransactionHolder.

ay0o5h reviewed (changes requested):
please check the changes

YoussefmSaber committed:
changed the directory name to datasource

YoussefmSaber committed:
added the delete and edit transaction methods to the interface

YoussefmSaber committed:
impelemented edit transaction function

YoussefmSaber committed:
impelemented delete transaction function

YoussefmSaber committed:
added comments to explain edit and delete transaction functions

YoussefmSaber committed:
commented the test cases cause it conflects with ci actions

YoussefmSaber commented:
## Changes that have been made: 
- removed the `saveTransactions()` function from the `DataSource` interface and it's implementation.
- refactored the `repository` directory and made it's name => `datasource` instead
- removed the `InMemoryTransactionStorage` and `TransactionHolder` files.
- renamed `loadTransactions()` function to => `getAllTransactions()`.
- added the `editTransaction()` and `deleteTransaction()` functions to the `DataSource` interface.
- implemented the `editTransaction()` and `deleteTransaction()` functions in the `FileTransactionStorage()` class.

## the new file structure
![image](https://github.com/user-attachments/assets/a287778a-96d7-469d-a236-dcff5bcaca0e)
